streaming/receiver.c

```diff
@@ -51,47 +51,235 @@ static void rrdpush_receiver_thread_cleanup(void *ptr) {
+#else
+    ret = send(host->receiver->fd, message, strlen(message), MSG_DONTWAIT);
+#endif
+    if (ret != (int)strlen(message))
+        error("Failed to send replication request for %s!", st_id);
+}
+
+static void skip_gap(RRDSET *st, time_t first_t, time_t last_t) {
+    // dbengine should handle time jump
+    if (st->rrd_memory_mode != RRD_MEMORY_MODE_DBENGINE) {
+        rrdset_rdlock(st);
+        RRDDIM *rd;
+        rrddim_foreach_read(rd, st) {
+            long current_entry = st->current_entry;
+            debug(D_REPLICATION, "Legacy-mode on %s.%s filling empty slots %ld-%ld.",
+                  st->name, rd->name, (long)first_t, (long)last_t);
+            for (time_t gap_t = first_t; gap_t <= last_t; gap_t += st->update_every) {
+                rd->values[current_entry] = SN_EMPTY_SLOT;
+                current_entry = ((current_entry + 1) >= st->entries) ? 0 : current_entry + 1;
+            }
         }
-        char message[128];
-        sprintf(message,"REPLICATE %ld %ld\n", remote_time - gap, remote_time);
-        int ret;
-#ifdef ENABLE_HTTPS
-        SSL *conn = host->stream_ssl.conn ;
-        if(conn && !host->stream_ssl.flags) {
-            ret = SSL_write(conn, message, strlen(message));
+        rrdset_unlock(st);
+    }
+    st->last_updated.tv_sec = last_t;
+}
+
+
+PARSER_RC streaming_rep_begin(char **words, void *user_v, PLUGINSD_ACTION *plugins_action) {
+    PARSER_USER_OBJECT *user = user_v;
+    UNUSED(plugins_action);
+    char *id = words[1];
+    char *start_txt = words[2];
+    char *first_txt = words[3];
+    char *end_txt = words[4];
+    if (!id || !start_txt || !end_txt)
+        goto disable;
+
+    RRDSET *st = NULL;
+    st = rrdset_find(user->host, id);
+    if (unlikely(!st))
+        goto disable;
+    user->st = st;
+    st->state->window_start = str2ll(start_txt, NULL);
+    st->state->window_first = str2ll(first_txt, NULL);
+    st->state->window_end   = str2ll(end_txt, NULL);
+
+    struct receiver_state *rpt = user->opaque;
+    time_t now = now_realtime_sec();
+    if (st->last_updated.tv_sec == 0) {
+        if (rpt->gap_history==0) {
+            st->last_updated.tv_sec = st->state->window_start;
+            debug(D_REPLICATION, "Initial data for %s, no historical gaps. window=%ld..%ld",
+                  st->name, (long)st->state->window_start, (long)st->state->window_end);
+            st->state->ignore_block = 0;
+            return PARSER_RC_OK;
         } else {
-            ret = send(host->receiver->fd, message, strlen(message), MSG_DONTWAIT);
+            debug(D_REPLICATION, "Initial data for %s, asking for history window=%ld..%ld",
+                  st->name, (long)now - rpt->gap_history+1, now);
+            send_replication_req(user->host, st->id, now - rpt->gap_history + 1, now);
+            st->state->ignore_block = 1;
+            st->last_updated.tv_sec = now - rpt->gap_history;
+            return PARSER_RC_OK;
```

Comment:
To complement what I wrote above in the previous function, I agree that when we reach this point the PARSER happened, but there was a problem to send data.

Reply:
If the data was not sent then it is retried the next the event happens - the overall design of the sender is idempotent (with respect to the changing state of the window). The return code indicates that the message was correctly parsed and the connection should be left open (if the parse is invalid then the stream is aborted).